docs/src/appendix/mast.md

```diff
@@ -0,0 +1,528 @@
+    ///
+    /// NOTE: The entrypoint is also present in `roots` if set
+    entrypoint: Option<MastNodeId>,
+}
+```
+
+This representation is "flattened", i.e. the nodes of the tree are stored in a vector, and edges between nodes are implied by referencing a node by its index in that vector. This will be more clear as we fill out the types mentioned above.
+
+This representation allows constant-time access to any node in the tree, as well as constant-time appends/insertions. Additionally, because all of the nodes are stored contiguously; and generally stored such that nodes in close proximity in the tree, are also close in the vector - accessing nearby nodes in the tree benefits from those nodes frequently being in the same cache line, or in a nearby cache line. It also has the effect of dramatically simplifying serialization/deserialization.
+
+The downside is that the tree structure is no longer enforced by the type system, instead we must maintain those invariants ourselves. This is a tradeoff, but one we've determined is worth making.
+
+Next, let's take a closer look at the definition of `MastNode` and the handle type used to reference `MastNode`s associated with a specific instance of `MastForest`:
+
+```rust
+/// We're using this simple type alias to reflect the fact that `MastNodeId`
+/// is ultimately an index into the `nodes` vector of a `MastForest`. However,
+/// in practice this is a more complex type that provides us with some
+/// additional guarantees/safety, and guards against accidental misuse.
+pub type MastNodeId = usize;
+
+pub enum MastNode {
+    /// See note following this snippet
+    Block(BasicBlockNode),
+    Join(JoinNode),
+    Split(SplitNode),
+    Loop(LoopNode),
+    Call(CallNode),
+    Dyn(DynNode),
+    /// A reference to a node whose definition is not
+    /// local to the containing `MastForest`.
+    External(Digest),
+}
+```
+
+> [!NOTE]
+> I'm using the "basic block" nomenclature here for clarity, and to better reflect its relationship to the semantics of basic blocks as they are known in standard compiler IRs and assemblers.
+>
+> In the current `miden-core` code, "block" is used for many things that are not actually blocks, or even block-like. This in turn seems to have forced the use of "span" to represent basic blocks. Since we've been starting to favor a more graph-theoretic set of terms for MAST in recent discussions, e.g. "tree"/"forest"/"node", I think it is clearer if we discard the previous usage of "block" in favor of "node", and "span" in favor of "basic block".
+
+The definitions of the concrete node types largely match what exists today, but we will look at them anyway just to illustrate any differences.
+
+First, a few preliminaries. We reference two traits which currently do not exist, but which represent common functionality across node types. We assume that they are derivable for purposes of this document:
+
+```rust
+pub trait MerkleTreeNode {
```

Comment:
Kind of - in our other discussions, `MastTreeNode` is a concrete enum/struct containing all of the node variants. `MerkleTreeNode` was just my way of defining a trait that conceptually is implemented by all `MastTreeNode` types. Since there is a decent chance both things would be needed, there is a bit of a naming conflict there, but we could obviously come up with some options. In any case, that's where this name came from, but I'm not sold on it.